extensions/amp-story/0.1/media-pool.js

```diff
@@ -0,0 +1,748 @@
+        return null;
+      }
+    }
+
+    // De-allocate a media element.
+    return allocatedEls.pop();
+  }
+
+
+  /**
+   * Evicts an element of the specified type, replaces it in the DOM with the
+   * original media element, and returns it.
+   * @param {!MediaType} mediaType The type of media element to evict.
+   * @param {!HTMLMediaElement=} opt_elToAllocate If specified, the element that
+   *     is trying to be allocated, such that another element must be evicted.
+   * @return {?HTMLMediaElement} A media element of the specified type.
+   * @private
+   */
+  evictMediaElement_(mediaType, opt_elToAllocate) {
+    const poolMediaEl =
+        this.deallocateMediaElement_(mediaType, opt_elToAllocate);
+    if (!poolMediaEl) {
+      return null;
+    }
+
+    this.swapPoolMediaElementOutOfDom_(poolMediaEl);
+    return poolMediaEl;
+  }
+
+
+  /**
+   * @param {!MediaType} mediaType The media type to check.
+   * @param {!HTMLMediaElement} el The element to check.
+   * @return {boolean} true, if the specified element has already been allocated
+   *     as the specified type of media element.
+   * @private
+   */
+  isAllocatedMediaElement_(mediaType, el) {
+    return this.allocated[mediaType].indexOf(el) >= 0;
+  }
+
+
+  /**
+   * Determines whether a CSS class name is allowed to be removed or copied from
+   * media elements.
+   * @param {string} cssClassName The CSS class name name to check.
+   * @return {boolean} true, if the specified CSS class name is allowed to be
+   *     removed or copied from media elements; false otherwise.
+   * @private
+   */
+  isProtectedCssClassName_(cssClassName) {
+    return PROTECTED_CSS_CLASS_NAMES.indexOf(cssClassName) >= 0;
+  }
+
+
+  /**
+   * Determines whether an attribute is allowed to be removed or copied from
+   * media elements.
+   * @param {string} attributeName The attribute name to check.
+   * @return {boolean} true, if the specified attribute is allowed to be removed
+   *     or copied from media elements; false otherwise.
+   * @private
+   */
+  isProtectedAttributeName_(attributeName) {
+    return PROTECTED_ATTRIBUTES.indexOf(attributeName) >= 0;
+  }
+
+
+  /**
+   * Replaces a media element that was originally in the DOM with a media
+   * element from the pool.
+   * @param {!HTMLMediaElement} domMediaEl The media element originating from
+   *     the DOM.
+   * @param {!HTMLMediaElement} poolMediaEl The media element originating from
+   *     the pool.
+   * @param {!Sources} sources The sources for the media element.
+   * @private
+   */
+  swapPoolMediaElementIntoDom_(domMediaEl, poolMediaEl, sources) {
+    this.copyCssClasses_(domMediaEl, poolMediaEl);
+    this.copyAttributes_(domMediaEl, poolMediaEl);
+    sources.applyToElement(poolMediaEl);
+    poolMediaEl.setAttribute(REPLACED_MEDIA_ATTRIBUTE, domMediaEl.id);
+    domMediaEl.parentElement.replaceChild(poolMediaEl, domMediaEl);
```

Comment:
without resetting the private `this.video_` reference in `amp-video`, this will break some functionality. Most may be irrelevant to STAMP but one that probably matters is "SessionMedia" stuff (e.g. when you play a video and then lock your phone, the lock screen shows the poster and media controls for the video that was just playing).

Also not setting `this.video_` will keep reference count > 0 which will keep the old dom node in memory for no good reason.

(I don't mind STAMP accessing .video_ for now, later however we may wanna to officially add `swapVideo()` to AMP core)

Let's also make sure there are enough tests in STAMPs to catch bug if `amp-video` does refactoring in a way that may break STAMP.

A good long-term solution could be STAMP having its own player `amp-stamp-video`. `amp-video` is not a lot of code anyway. Most things are handle by `video-manager` which STAMP version of video player can register to as well.

Reply:
`MediaPool` actually keeps the DOM node in memory anyway; if a video is evicted from the `MediaPool`, we put back the original video.

Reply:
I see. I assume simply unattaching it from the DOM frees up hardware decoding space on iOS though